### ISSUE TITLE
How to handle derived types in MediatR Notifications

### ISSUE THREAD

**Issue by hannahchan** [jbogard/MediatR#678]
Hi,

I'm hoping to learn something here. I'm currently using MediatR Notifications as an in-process event bus in an ASP.NET Core app with the default IoC container.

I've structured my events (MediatR Notifications) with the following inheritance chain.

```
INotification, IAuditableEvent -> ProductEvent -> ProductAddedEvent
INotification, IAuditableEvent -> ProductEvent -> ProductUpdatedEvent
INotification, IAuditableEvent -> ProductEvent -> ProductRemovedEvent

INotification, IAuditableEvent -> VendorEvent -> VendorAddedEvent
INotification, IAuditableEvent -> VendorEvent -> VendorUpdatedEvent
INotification, IAuditableEvent -> VendorEvent -> VendorRemovedEvent
```

ProductEvent and VendorEvent are abstract classes.

Writing a notification handler is pretty straight forward when you only want handle the most derived type of an event.

```C#
public class ProductAddedEventHandler : INotificationHandler<ProductAddedEvent>
{
    public Task Handle(ProductAddedEvent notification, CancellationToken cancellationToken)
    {
        // Event handling code goes here
        return Task.CompletedTask;
    }
}
```

However sometimes I want a handler that handles a lesser derived type.

```C#
public class ProductEventHandler : INotificationHandler<ProductEvent>
{
    public Task Handle(ProductEvent notification, CancellationToken cancellationToken)
    {
        // Event handling code goes here
        return Task.CompletedTask;
    }
}
```

The above handler never seems to trigger when handling the more derived types. I've register my handlers with `services.AddMediatR(typeof(Startup).Assembly)` and am wondering if there is a gap in my knowledge on how the default IoC container works or if it is something else entirely different.

My current workaround is below.

```C#
public class ProductEventHandler<TNotification> : INotificationHandler<TNotification>
    where TNotification : INotification
{

    public Task Handle(TNotification notification, CancellationToken cancellationToken)
    {
        if (notification is ProductEvent productEvent)
        {
            // Event handling code goes here
        }

        return Task.CompletedTask;
    }
}
```
What are the alternatives to what I have done?

**Comment by BGajda13** [comment]
Try that:
```
public class ProductEventHandler<TNotification> : INotificationHandler<TNotification>
    where TNotification : ProductEvent
{

    public Task Handle(TNotification notification, CancellationToken cancellationToken)
    {
        return Task.CompletedTask;
    }
}
```
But it will fire for every type which inherits from ProductEvent

**Comment by hannahchan** [comment]
Thanks @BGajda13. This works!

I swear I did try this earlier and it didn't work when I had two handlers, one for `ProductEvents` and one for `VendorEvents`. The `VendorEvent` one kept getting invoked when I published a `ProductEvent` and then would crash during runtime. I thought it would be the way I was registering the handlers.

I might go off and do some more testing. I tried your suggested code in a .NET 6 application and it works. I suspect this doesn't work in a .NET Core 2.1 application which is what I was working on.

**Comment by hannahchan** [comment]
I just tested this again in .NET 6 with MediatR 9.0.0 and it appears to work.

When I tested this in an ASP.NET Core 2.1 app with MediatR 9.0.0 and I get the following exception;

```
System.ArgumentException: GenericArguments[0], 'Messaging.ProductsUpdatedEvent', on 'Handlers.VendorEventHandler`1[TNotification]' violates the constraint of type 'TNotification'. ---> System.TypeLoadException: GenericArguments[0], 'Messaging.ProductsUpdatedEvent', on 'Handlers.VendorEventHandler`1[TNotification]' violates the constraint of type parameter 'TNotification'.
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware:Error: An unhandled exception has occurred while executing the request.

System.ArgumentException: GenericArguments[0], 'Messaging.ProductsUpdatedEvent', on 'Handlers.VendorEventHandler`1[TNotification]' violates the constraint of type 'TNotification'. ---> System.TypeLoadException: GenericArguments[0], 'Messaging.ProductsUpdatedEvent', on 'Handlers.VendorEventHandler`1[TNotification]' violates the constraint of type parameter 'TNotification'.
   at System.RuntimeTypeHandle.Instantiate(RuntimeTypeHandle handle, IntPtr* pInst, Int32 numGenericArgs, ObjectHandleOnStack type)
   at System.RuntimeTypeHandle.Instantiate(Type[] inst)
   at System.RuntimeType.MakeGenericType(Type[] instantiation)
   --- End of inner exception stack trace ---
   at System.RuntimeType.ValidateGenericArguments(MemberInfo definition, RuntimeType[] genericArguments, Exception e)
   at System.RuntimeType.MakeGenericType(Type[] instantiation)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateOpenGeneric(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateEnumerable(Type serviceType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateCallSite(Type serviceType, CallSiteChain callSiteChain)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.CreateServiceAccessor(Type serviceType)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at MediatR.ServiceFactoryExtensions.GetInstances[T](ServiceFactory factory)
   at MediatR.Internal.NotificationHandlerWrapperImpl`1.Handle(INotification notification, CancellationToken cancellationToken, ServiceFactory serviceFactory, Func`4 publish)
   at MediatR.Mediator.PublishNotification(INotification notification, CancellationToken cancellationToken)
   at MediatR.Mediator.Publish[TNotification](TNotification notification, CancellationToken cancellationToken)
   at EDG.CCE.API.Controllers.Internal.ProductsController.Patch(Int64 id, JsonPatchDocument`1 patchProduct) in C:\Users\1162130\Code\CONNECTIONS-API\EDG.CCE.API\Controllers\Internal\ProductsController.cs:line 263
   at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

I guess I just found another reason to move off .NET Core 2.1 other than it being no longer supported.

**Comment by BGajda13** [comment]
@hannahchan I think your problem is highly coupled with IoC Container. 2.1 can have all goods implemented, as 6.0 have. Generally, I'm speaking about polymorphic dispatch, which is highly dependent on IoC Container. See [here](https://github.com/jbogard/MediatR/wiki#polymorphic-dispatch).

**Comment by hannahchan** [comment]
Yep this is definitely an IoC Container issue. I'm more confident now that we can close this knowing that upgrading to the latest version of .NET solves this.
